channeldb/payments.go

```diff
@@ -383,6 +383,158 @@ func fetchHtlcFailInfo(bucket *bbolt.Bucket) (*HTLCFailInfo, error) {
+// and at the previous index for reversed order.
+// startIndex returns a boolean, which tells about whether the indexOffset
+// would give rise to payments found in the next when looping through them.
+func startIndex(indexOffset, paymentsLength uint64, reversed bool) (uint64, bool) {
+	if reversed {
+		switch {
+		// When indexOffset is zero or the indexOffset lies outside
+		// of the payments range, we start from the last known payment.
+		case indexOffset == 0 || indexOffset > paymentsLength:
+			return paymentsLength, true
+		case indexOffset > 1:
+			return indexOffset - 1, true
+		default:
+			return 0, false
+		}
+	}
+	// Handle the normal (forwards) order.
+	switch {
+	case indexOffset < paymentsLength:
+		return indexOffset + 1, true
+	default:
+		return 0, false
+	}
+}
+
+// nextIndex determines the next payment index in the loop.
+// It includes a boolean which tells about whether there was a next index.
+func nextIndex(index, paymentsLength uint64, reversed bool) (uint64, bool) {
+	switch {
+	case reversed && index > 1:
+		return index - 1, true
+	case !reversed && index < paymentsLength:
+		return index + 1, true
+	// Stop if payments are exhausted.
+	default:
+		return 0, false
+	}
+}
+
+// QueryPayments is a query to the payments database which is restricted
+// to a subset of payments by the payments query, containing an offset
+// index and a maximum number of returned payments.
+func (db *DB) QueryPayments(query PaymentsQuery) (PaymentsQuerySlice, error) {
+	resp := PaymentsQuerySlice{
+		PaymentsQuery: query,
+	}
+	allPayments, err := db.FetchPayments()
+	if err != nil {
+		return resp, err
+	}
+	paymentsLength := uint64(len(allPayments))
+
+	if paymentsLength == 0 {
+		return resp, nil
+	}
+
+	i, ok := startIndex(query.IndexOffset, paymentsLength, query.Reversed)
+
+	var count uint64
+	for ; count < query.NumMaxPayments && ok; i, ok = nextIndex(
+		i, paymentsLength, query.Reversed) {
+		// To keep compatibility with the old API, we only return
+		// non-succeeded payments if requested.
+		if allPayments[i-1].Status != StatusSucceeded &&
+			!query.IncludeIncomplete {
+			continue
+		}
+
+		// Update first and last index offsets for the set of returned
+		// payments. If we traverse in the reversed order we remember
+		// the first handled payment and set the LastIndexOffset to its
+		// index. The FirstIndexOffset is updated every time we add
+		// another payment. In the forward order we only set the
+		// FirstIndexOffset once and update the LastIndexOffset.
+		if query.Reversed {
+			resp.FirstIndexOffset = i
+			if resp.LastIndexOffset == 0 {
+				resp.LastIndexOffset = i
+			}
+		} else {
+			if resp.FirstIndexOffset == 0 {
+				resp.FirstIndexOffset = i
+			}
+			resp.LastIndexOffset = i
+		}
+
+		resp.Payments = append(resp.Payments, *allPayments[i-1])
+		count++
+	}
+
+	// Need to swap the payments slice order if reversed order.
+	if query.Reversed {
```

Comment:
Why swap actually? So the idea is that the pages are reversed, but per page the payments are in ascending order? Also something to document in the proto.

Reply:
It's again a replication of `listinvoices` behavior. Have documented in proto under the `reversed` flag documentation.